src/lib/server/endpoints/images.ts

```diff
@@ -0,0 +1,214 @@
+import type { Sharp } from "sharp";
+import sharp from "sharp";
+import type { MessageFile } from "$lib/types/Message";
+import { z } from "zod";
+
+export function createImageProcessorOptionsValidator<MimeType extends string = string>(
+	defaults: ImageProcessorOptions<MimeType>
+) {
+	return z
+		.object({
+			supportedMimeTypes: z
+				.array(
+					z.enum<string, [MimeType, ...MimeType[]]>([
+						defaults.supportedMimeTypes[0],
+						...defaults.supportedMimeTypes.slice(1),
+					])
+				)
+				.default(defaults.supportedMimeTypes),
+			preferredMimeType: z
+				.enum<string, [MimeType, ...MimeType[]]>([
+					defaults.supportedMimeTypes[0],
+					...defaults.supportedMimeTypes.slice(1),
+				])
+				.default(defaults.preferredMimeType),
+			// The 4 / 3 compensates for the 33% increase in size when converting to base64
+			maxSizeInMB: z.number().positive().default(defaults.maxSizeInMB),
+			maxWidth: z.number().int().positive().default(defaults.maxWidth),
+			maxHeight: z.number().int().positive().default(defaults.maxHeight),
+		})
+		.default(defaults);
+}
+
+export interface ImageProcessorOptions<MimeType extends string = string> {
+	supportedMimeTypes: MimeType[];
+	preferredMimeType: MimeType;
+	maxSizeInMB: number;
+	maxWidth: number;
+	maxHeight: number;
+}
+export type ImageProcessor<MimeType extends string = string> = (file: MessageFile) => Promise<{
+	image: Buffer;
+	mime: MimeType;
+}>;
+export function makeImageProcessor<MimeType extends string = string>(
+	options: ImageProcessorOptions<MimeType>
+): ImageProcessor<MimeType> {
+	return async (file) => {
+		const { supportedMimeTypes, preferredMimeType, maxSizeInMB, maxWidth, maxHeight } = options;
+		const { mime, value } = file;
+
+		const buffer = Buffer.from(value, "base64");
+		let sharpInst = sharp(buffer);
+
+		const metadata = await sharpInst.metadata();
+		if (!metadata) throw Error("Failed to read image metadata");
+		const { width, height } = metadata;
+		if (width === undefined || height === undefined) throw Error("Failed to read image size");
+
+		const tooLargeInSize = width > maxWidth || height > maxHeight;
+		const tooLargeInBytes = buffer.byteLength > maxSizeInMB * 1024 * 1024;
```

Comment:
Is `* 1024 * 1024` correct here OR should be `* 1000 * 1000` ?

Reply:
good catch!

Reply:
btw, was this comment addressed? Which value is the correct one in the end?

Reply:
switched to 1000 x 1000 for safety